pkg/gather/clusterconfig/workloads.go

```diff
@@ -0,0 +1,559 @@
+	imageConfig.Burst = 10
+	gatherOpenShiftClient, err := imageclient.NewForConfig(imageConfig)
+	if err != nil {
+		c <- gatherResult{nil, []error{err}}
+		return
+	}
+	result, errs := gatherWorkloadInfo(g.ctx, gatherKubeClient.CoreV1(), gatherOpenShiftClient)
+	c <- gatherResult{result, errs}
+}
+
+func gatherWorkloadInfo(ctx context.Context, coreClient corev1client.CoreV1Interface, imageClient imageclient.ImageV1Interface) ([]record.Record, []error) {
+	// load images as we find them
+	imageCh := make(chan string, workloadGatherPageSize)
+	imagesDoneCh := gatherWorkloadImageInfo(ctx, imageClient.Images(), imageCh)
+
+	// load pods in order
+	start := time.Now()
+
+	var info workloadPods
+	var namespace string
+	var namespaceHash string
+	var namespacePods workloadNamespacePods
+	h := sha256.New()
+
+	// Use the Limit and Continue fields to request the pod information in chunks.
+	var continueValue string
+	for {
+		pods, err := coreClient.Pods("").List(ctx, metav1.ListOptions{
+			Limit:    workloadGatherPageSize,
+			Continue: continueValue,
+		})
+		if err != nil {
+			return nil, []error{err}
+		}
+
+		for _, pod := range pods.Items {
+			// initialize the running state, including the namespace hash
+			if pod.Namespace != namespace {
+				if len(namespace) != 0 {
+					if info.Namespaces == nil {
+						info.Namespaces = make(map[string]workloadNamespacePods, 1024)
+					}
+					info.Namespaces[namespaceHash] = namespacePods
+					info.PodCount += namespacePods.Count
+				}
+				namespace = pod.Namespace
+				namespaceHash = workloadHashString(h, namespace)
+				namespacePods = workloadNamespacePods{Shapes: make([]workloadPodShape, 0, 16)}
+			}
+
+			namespacePods.Count++
+
+			switch {
+			case pod.Status.Phase == corev1.PodSucceeded, pod.Status.Phase == corev1.PodFailed:
+				// track terminal pods but do not report their data
+				namespacePods.TerminalCount++
+				continue
+			case pod.Status.Phase != corev1.PodRunning && pod.Status.Phase != corev1.PodPending:
+				// only consider pods that are in a known state
+				namespacePods.IgnoredCount++
+				continue
+			case len(pod.Status.InitContainerStatuses) != len(pod.Spec.InitContainers), len(pod.Status.ContainerStatuses) != len(pod.Spec.Containers):
+				// pods without filled out status are invalid
+				namespacePods.IgnoredCount++
+				continue
+			}
+
+			var podShape workloadPodShape
+			var ok bool
+			podShape.InitContainers, ok = calculateWorkloadContainerShapes(h, pod.Spec.InitContainers, pod.Status.InitContainerStatuses)
+			if !ok {
+				namespacePods.InvalidCount++
+				continue
+			}
+			podShape.Containers, ok = calculateWorkloadContainerShapes(h, pod.Spec.Containers, pod.Status.ContainerStatuses)
+			if !ok {
+				namespacePods.InvalidCount++
+				continue
+			}
+
+			if index := workloadPodShapeIndex(namespacePods.Shapes, podShape); index != -1 {
+				namespacePods.Shapes[index].Duplicates++
+			} else {
+				namespacePods.Shapes = append(namespacePods.Shapes, podShape)
```

Comment:
Do we need the `containers` array to distinguish the pods (for each pod basically)? I mean, now I can get something like: 
```
....
                ,{
                      "containers": [
                        {
                            "imageID": "sha256:bb525a274fcc67b5a8f05882550a77e6586434cf9e78390f2632d6c558456fb2"
                        }
                    ]
                },
                {
                    "containers": [
                        {
                            "imageID": "sha256:3aeeb731c3254f5841cbd679e68ee409d1a1aaabc745c803ae2010f57dbe03cc",
                            "firstCommand": "zC8CUf1mHJ6g"
                        }
                    ]
                },
...
```
but maybe it could be only something like:
```
                {
                    "containers": [
                        {
                            "imageID": "sha256:bb525a274fcc67b5a8f05882550a77e6586434cf9e78390f2632d6c558456fb2"
                        },
                        {
                            "imageID": "sha256:3aeeb731c3254f5841cbd679e68ee409d1a1aaabc745c803ae2010f57dbe03cc",
                            "firstCommand": "zC8CUf1mHJ6g"
                        },
                    ]
```

Reply:
knowing the difference between two shapes is mostly so we can tell the difference between certain similar uses of images - the hypothesis is that .there is a fair amount of image reuse but it’s in different pods.  if we collapsed then we’d miss some examples

Reply:
OK thanks. I would suggest the following. @smarterclayton can you please remove the WIP? We would merge it then and added some limits (it will run less often + some limit for the number of pods - I would go with 2500), which we can always remove in the future.

Reply:
Yeah, let me profile this really quick a few more times to see what i missed.

Reply:
So for a very large (p95 of the fleet, roughly) cluster iwth 6000 pods, this used about 30mb memory peak (allocated 150mb total), took 15s, and used about 2 core seconds, which is effectively "free".  Size of gathered was 30k.  

I'm not sure there's a need to limit or restrict size - even on a 600k pod cluster, that's roughly still "free".

Reply:
I would probably say when we get to 30-50k pods we'd probably want to start limiting.  I can do a follow up which either runs less frequently or simply disables the capture at those levels

Reply:
Thanks for the analysis. Sounds good, but I am always a little careful. I mean I am not sure on the impact on our processing pipelines so I wanted to start low. Only 30k for the size of the data? It seems little....only the container sha256 is around 100 bytes

Reply:
I tested it on a cluster-bot cluster where I deployed 20,000 sleeping pods, so its not a perfect test case but my findings are similar to Clayton's.
The increase in CPU usage is nothing extreme (~ 5 core seconds).
The RAM got a bit big if we look at total alloc of 240MB just for this, but I didn't see any major peak in memory usage. 
It took 22 seconds. (The size of the archive didn't change significantly, but that is because the 20,000 duplicates are not stored)
Its definitely the "heaviest" gathering we have currently (on large clusters), but it shouldn't cause trouble.